qemu/cpu/code_generation.py

```diff
@@ -528,12 +530,70 @@ def fill_env_get_cpu_body(cputype, function):
         )
     )
 
-def fill_gdb_rw_register_body(cputype, function, comment):
+gdb_get_reg_bitsizes = (8, 16, 32, 64, 128)
+ld_size_infixes = {16: "uw", 32: "l", 64: "q"}
+
+def fill_gdb_rw_register_body(cputype, function, is_write = False):
     cpu = Pointer(Type[cputype.struct_instance_name])("cpu")
     env = Pointer(Type[cputype.struct_name])("env")
 
+    n = function.args[2]
+
+    reg_number = 0
+    cases = []
+    for reg in cputype.registers:
+        bank_size = reg.bank_size
+        raw_bitsize = reg.raw_bitsize
+        env_reg = OpSDeref(env, reg.name)
+
+        if bank_size:
+            env_reg = OpIndex(
+                env_reg,
+                OpSub(n, reg_number) if reg_number else n
+            )
+            case_val = CaseRange(reg_number, reg_number + bank_size - 1)
+            reg_number += bank_size
+        else:
+            case_val = reg_number
+            reg_number += 1
+
+        if is_write:
+            ld_bitsize, ld_size_infix = min(filter(
+                lambda infix: infix[0] >= raw_bitsize,
+                ld_size_infixes.items()
+            ))
+            ld_call = Call( "ld%s_p" % (ld_size_infix), function.args[1])
+            if raw_bitsize != ld_bitsize:
+                ld_call = OpAnd(
+                    ld_call,
+                    CINT((1 << raw_bitsize) - 1, base = 16)
+                )
+            cases.append(
+                SwitchCase(case_val, add_break = False)(
+                    OpAssign(env_reg, ld_call),
+                    Return(ld_bitsize // BYTE_BITSIZE)

```

Comment:
We read `ld_bitsize//BYTE_BITSIZE` from buffer (because of only such a helper is available) but we actually used `(raw_bitsize + BYTE_BITSIZE - 1) // BYTE_BITSIZE`.
Should we `return (raw_bitsize + BYTE_BITSIZE - 1) // BYTE_BITSIZE`?

Reply:
MSP430X has 20bit registers. And I don't remember if GDB RSP uses 4 or 3 bytes to pack them.